File: LeetCode/4.median-of-two-sorted-arrays.py
#
# @lc app=leetcode id=4 lang=python3
#
# [4] Median of Two Sorted Arrays
#
# https://leetcode.com/problems/median-of-two-sorted-arrays/description/
#
# algorithms
# Hard (43.24%)
# Likes:    29819
# Dislikes: 3354
# Total Accepted:    3.3M
# Total Submissions: 7.6M
# Testcase Example:  '[1,3]\n[2]'
#
# Given two sorted arrays nums1 and nums2 of size m and n respectively, return
# the median of the two sorted arrays.
# 
# The overall run time complexity should be O(log (m+n)).
# 
# 
# Example 1:
# 
# 
# Input: nums1 = [1,3], nums2 = [2]
# Output: 2.00000
# Explanation: merged array = [1,2,3] and median is 2.
# 
# 
# Example 2:
# 
# 
# Input: nums1 = [1,2], nums2 = [3,4]
# Output: 2.50000
# Explanation: merged array = [1,2,3,4] and median is (2 + 3) / 2 = 2.5.
# 
# 
# 
# Constraints:
# 
# 
# nums1.length == m
# nums2.length == n
# 0 <= m <= 1000
# 0 <= n <= 1000
# 1 <= m + n <= 2000
# -10^6 <= nums1[i], nums2[i] <= 10^6
# 
# 
#

# @lc code=start
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        # nums = nums1 + nums2
        # nums.sort()
        # n = len(nums)
        # if n % 2 == 0:
        #     return (nums[n // 2 - 1] + nums[n // 2]) / 2
        # else:
        #     return nums[n // 2]


        # len1, len2 = len(nums1), len(nums2)
        # total_len = len1 + len2
        # mid = total_len // 2

        # i, j = 0, 0  # 分别表示 nums1 和 nums2 的指针
        # med1, med2 = 0, 0

        # for _ in range(mid + 1):
        #     med1 = med2
        #     if i < len1 and (j >= len2 or nums1[i] < nums2[j]):
        #         med2 = nums1[i]
        #         i += 1
        #     else:
        #         med2 = nums2[j]
        #         j += 1

        # if total_len % 2 == 0:
        #     return (med1 + med2) / 2
        # else:
        #     return med2
        
        
        def find_kth_element(k):
            index1, index2 = 0, 0  # 分别表示 nums1 和 nums2 的起始索引
            while True:
                # 边界情况
                if index1 == len(nums1):  # nums1 已经耗尽
                    return nums2[index2 + k - 1]
                if index2 == len(nums2):  # nums2 已经耗尽
                    return nums1[index1 + k - 1]
                if k == 1:  # 找到第 1 小的元素
                    return min(nums1[index1], nums2[index2])

                # 正常情况：比较两个数组的第 k//2 个元素
                new_index1 = min(index1 + k // 2 - 1, len(nums1) - 1)
                new_index2 = min(index2 + k // 2 - 1, len(nums2) - 1)
                pivot1, pivot2 = nums1[new_index1], nums2[new_index2]

                if pivot1 <= pivot2:
                    k -= (new_index1 - index1 + 1)
                    index1 = new_index1 + 1
                else:
                    k -= (new_index2 - index2 + 1)
                    index2 = new_index2 + 1

        total_len = len(nums1) + len(nums2)
        if total_len % 2 == 1:  # 奇数长度
            return find_kth_element(total_len // 2 + 1)
        else:  # 偶数长度
            return (find_kth_element(total_len // 2) + find_kth_element(total_len // 2 + 1)) / 2
        
# @lc code=end

File: 00LeetCodeCpp/4.median-of-two-sorted-arrays.cpp
/*
 * @lc app=leetcode id=4 lang=cpp
 *
 * [4] Median of Two Sorted Arrays
 */

// @lc code=start
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size();
        int m = nums2.size();

        int left = (n + m + 1) / 2;  // 左中位数位置
        int right = (n + m + 2) / 2; // 右中位数位置
        
        // 技巧：如果是奇数，left == right，逻辑统一；如果是偶数，求两个位置的平均值
        return (getKth(nums1, 0, n - 1, nums2, 0, m - 1, left) + 
                getKth(nums1, 0, n - 1, nums2, 0, m - 1, right)) * 0.5;
    }
private:
    // 在两个数组的指定区间内，寻找第 k 小的数
    int getKth(vector<int>& nums1, int start1, int end1, vector<int>& nums2, int start2, int end2, int k) {
        int len1 = end1 - start1 + 1;
        int len2 = end2 - start2 + 1;

        // 始终让 nums1 成为较短的数组，方便处理越界情况
        if (len1 > len2) return getKth(nums2, start2, end2, nums1, start1, end1, k);
        
        // 边界情况 1：短数组已经排空了，直接去长数组找
        if (len1 == 0) return nums2[start2 + k - 1];
        
        // 边界情况 2：找第 1 小的数，直接比两个数组的头，谁小就是谁
        if (k == 1) return min(nums1[start1], nums2[start2]);

        // 核心逻辑：二分，各自取第 k/2 个数进行比较
        // 注意：为了防止短数组长度不足 k/2，取 min(len1, k/2)
        int i = start1 + min(len1, k / 2) - 1;
        int j = start2 + min(len2, k / 2) - 1;

        if (nums1[i] > nums2[j]) {
            // nums2 的前半部分（j 之前）绝对不可能是第 k 小，砍掉！
            // 更新 k：减去被砍掉的长度
            return getKth(nums1, start1, end1, nums2, j + 1, end2, k - (j - start2 + 1));
        } else {
            // nums1 的前半部分（i 之前）绝对不可能是第 k 小，砍掉！
            return getKth(nums1, i + 1, end1, nums2, start2, end2, k - (i - start1 + 1));
        }
    }
};
// @lc code=end

